### DISCUSSION TITLE
On testing

### DISCUSSION THREAD

**Discussion by amuradyan** [amuradyan/sea#2] [Ideas]
Roughly, this:

```scheme
(module Adder
  (export add/2)
  
  (define (add x y)
    (test (add 0 1) (= actual 1))
    (test (add 1 0) (= actual 1))
    (test (add -1 1) (= actual 0))
    (+ x y)))

(module Main
  (import (Adder add/2))
  
  (define (main)
    (display (add 1 "2")) ; should not compile, since no tests are found
    (display (add 1 2)) ; 3
    (display (add 0 0)))) ; 0
```

As I see it now, the only thing `test` does is it evaluates its first argument, binds it to the `actual` and passes it to its secon argument.

What about PBT?

**Comment by amuradyan** [comment]
One issue with the example above is that it does not provide a _readable_ result. Maybe we can try something like this?

```scheme
  (define (add x y)
    (test ("that zero does not change the sum" (add 0 1)) (= actual 1))
    (test ("that zero does not change the sum" (add 1 0)) (= actual 1))
    (test ("that zero does not change the sum" (add 0 0)) (= actual 0))
    (test ("that inverses should add up to zero" (add -1 1)) (= actual 0))
    (test ("that inverses should add up to zero" (add -2 2)) (= actual 0))
    (+ x y))
```

We can even _sugarcoat_ it like so. It seems this can be easily achieved with macros. Not sure if it is worth it though.

```scheme
(test ("that zero does not change the sum" for 0 1) (= actual 1))
```

**Comment by amuradyan** [comment]
It should be nice to design the `test` to be used independently, just as any other method, and expect it to provide a visually suggestive /red, green/ output. Unfortunately, GitHub does not allow colored text in comments, so I'll paste an image instead.

![image](https://user-images.githubusercontent.com/9533055/229248933-b83be5fa-93b4-4da2-b03f-9b5bff7c5429.png)

**Reply by antaranyan** [reply]
ստեղ /նորմալ/ դիֆֆ ա ուզվում
